### PR TITLE
Add dry-run mode

### PR DESCRIPTION
**Pull request by stephenmcgruer**

### CODE REVIEW

File: lib/comment.js [hunk 1]
```diff
@@ -20,26 +20,37 @@ function isExpeditedCss(metadata) {
...
     if (metadata.reviewedDownstream) {
-        return approve(
-            number,
-            'The review process for this patch is being conducted in the ' +
-                metadata.reviewedDownstream + ' project.'
-        );
+        const message = 'The review process for this patch is being conducted in the ' +
+            metadata.reviewedDownstream + ' project.';
+        if (dryRun) {
+            console.log(`[DRY-RUN] Would approve issue ${number} with message: ${message}`);
```
Comment (zcorpan) [inline lib/comment.js]:
```suggestion
            console.log(`[DRY-RUN] Would approve PR ${number} with message: ${message}`);
```

File: lib/comment.js [hunk 2]
```diff
@@ -20,26 +20,37 @@ function isExpeditedCss(metadata) {
...
     if (isExpeditedCss(metadata)) {
-        return approve(
-            number,
-            '[Peer review is not required for CSS editors.]' +
-                '(https://www.w3.org/2018/10/22-css-minutes.html#item20)'
-        );
+        const message = '[Peer review is not required for CSS editors.]' +
+            '(https://www.w3.org/2018/10/22-css-minutes.html#item20)';
+        if (dryRun) {
+            console.log(`[DRY-RUN] Would approve issue ${number} with message: ${message}`);
```
Comment (zcorpan) [inline lib/comment.js]:
```suggestion
            console.log(`[DRY-RUN] Would approve PR ${number} with message: ${message}`);
```

File: lib/comment.js [hunk 3]
```diff
@@ -20,26 +20,37 @@ function isExpeditedCss(metadata) {
...
+            return Promise.resolve(true);
+        }
+
+        return approve(number, message);
     }
 
     return promise().then(function() {
         if (metadata.missingAssignee) {
+            if (dryRun) {
+                console.log(`[DRY-RUN] Would add assignee ${metadata.missingAssignee} to issue ${number}`);
```
Comment (zcorpan) [inline lib/comment.js]:
```suggestion
                console.log(`[DRY-RUN] Would add assignee ${metadata.missingAssignee} to PR ${number}`);
```

File: lib/comment.js [hunk 4]
```diff
@@ -48,6 +59,15 @@ module.exports = function(number, metadata) {
         // if there are collaborators, we ask them for a review
         if (metadata.reviewersExcludingAuthor.length > 0 || metadata.isRoot) {
             if (metadata.missingReviewers.length > 0) {
+                if (dryRun) {
+                    console.log(`[DRY-RUN] Would add reviewers ${metadata.missingReviewers} to issue ${number}`);
```
Comment (zcorpan) [inline lib/comment.js]:
```suggestion
                    console.log(`[DRY-RUN] Would add reviewers ${metadata.missingReviewers} to PR ${number}`);
```

File: lib/comment.js [hunk 5]
```diff
@@ -48,6 +59,15 @@ module.exports = function(number, metadata) {
...
+                    // Assume that adding reviewers would have succeeded.
+                    if (!metadata.isMergeable) {
+                        console.log(`[DRY-RUN] Would add label "status:needs-collaborator" and comment "No reviewer on this pull request has write-access to the repository." to issue ${number}`);
```
Comment (zcorpan) [inline lib/comment.js]: Put the label and message in variables, and change "issue" to "PR".

Similarly below.

File: lib/label-model.js [hunk 6]
```diff
@@ -4,11 +4,16 @@ var github = require('./github'),
...
 
 exports.post = post;
-function post(issue, labels) {
+function post(issue, labels, dryRun) {
     return get(issue).then(function(existingLabels) {
         return complement(existingLabels, labels);
     }).then(function(labels) {
         if (!labels.length) { return labels; }
+        if (dryRun) {
+            console.log(`[DRY-RUN] Would add labels ${labels} to issue ${issue}`);
```
Comment (zcorpan) [inline lib/label-model.js]:
```suggestion
            console.log(`[DRY-RUN] Would add labels ${labels} to PR ${issue}`);
```

File: lib/comment.js [hunk 7]
```diff
@@ -48,30 +59,60 @@ module.exports = function(number, metadata) {
...
+                }
+                return labelAndComment(number, label, message);
             }
         } else {
             if (metadata.author.isOwner) {
-                return labelAndComment(number, "status:needs-reviewers", "There are no reviewers for this pull request besides its author.");
+                const label = "status:needs-reviewers";
+                const message = "There are no reviewers for this pull request besides its author.";
+                if (dryRun) {
+                    console.log(`[DRY-RUN] Would add label "${label} and comment "${message}" to PR ${number}`);
```
Comment (zcorpan) [inline lib/comment.js]:
```suggestion
                    console.log(`[DRY-RUN] Would add label "${label}" and comment "${message}" to PR ${number}`);
```

**Comment by zcorpan** [comment]
Oh, this was merging into `appengine`, I missed that. This change is not deployed on heroku.

**Comment by stephenmcgruer** [comment]
That is expected and desired, thanks :). The plan is to try and get an instance running in `dry-run` mode on appengine before we go any further.

**Pull request merged**